app/components/app_patient_programmes_table_component.rb

```diff
@@ -25,63 +25,143 @@ def call
   HEADERS = ["Programme name", "Status", "Notes"].freeze
 
   def rows
-    programmes.flat_map { |programme| rows_for_programme(programme) }
+    programmes.flat_map { |programme| rows_for_programme(programme:) }
   end
 
-  def rows_for_programme(programme)
+  def rows_for_programme(programme:)
     if programme.seasonal?
-      seasonal_programme_rows(programme)
+      seasonal_programme_rows(programme:)
     else
-      non_seasonal_programme_rows(programme)
+      non_seasonal_programme_rows(programme:)
     end
   end
 
-  def seasonal_programme_rows(programme)
+  def seasonal_programme_rows(programme:)
     eligible_year_groups = eligible_year_groups_for(programme:)
 
     AcademicYear.all.flat_map do |academic_year|
       year_group = patient.year_group(academic_year:)
       next unless year_group.in?(eligible_year_groups)
 
-      if vaccinated_for_academic_year?(programme:, academic_year:)
-        vaccination_records_for(programme).map do |vaccination_record|
-          build_row(programme, academic_year, vaccination_record)
+      vaccination_records = vaccination_records_for(programme:, academic_year:)
+
+      rows =
+        (1..vaccination_records.count - 1).to_a.map do |dose_sequence|
+          build_row(
+            programme:,
+            academic_year:,
+            vaccination_records: vaccination_records.slice(0, dose_sequence)
+          )
         end
-      else
-        [build_row(programme, academic_year)]
-      end
+
+      rows +
+        [
+          build_row_from_cache(
+            programme:,
+            academic_year:,
+            last_vaccination_record: vaccination_records.last
+          )
+        ]
     end
   end
 
-  def non_seasonal_programme_rows(programme)
+  def non_seasonal_programme_rows(programme:)
     academic_year = AcademicYear.pending
-    vaccination_records = vaccination_records_for(programme)
-
-    if vaccination_records.any?
-      vaccination_records.map do |vaccination_record|
-        build_row(programme, academic_year, vaccination_record)
+    vaccination_records = vaccination_records_for(programme:)
+
+    rows =
+      (1..vaccination_records.count - 1).to_a.map do |dose_sequence|
+        build_row(
+          programme:,
+          academic_year:,
+          vaccination_records: vaccination_records.slice(0, dose_sequence)
+        )
       end
-    else
-      [build_row(programme, academic_year)]
-    end
+
+    rows +
+      [
+        build_row_from_cache(
+          programme:,
+          academic_year:,
+          last_vaccination_record: vaccination_records.last
+        )
+      ]
   end
 
-  def build_row(programme, academic_year, vaccination_record = nil)
+  def build_row(programme:, academic_year:, vaccination_records: nil)
+    status =
+      StatusGenerator::Vaccination.new(
```

Comment:
Is there a way we can use the `Patient#vaccination_status` method to get these values? The generator classes weren't intended to be used outside the context of generating a status to be cached.

Reply:
Can we not then display the `outcome` of the vaccination record rather than needing to generate an overall status?